File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/ui/states/StateTimeAndDeviceIdFragment.java
package com.klnvch.greenhousecommon.ui.states;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.klnvch.greenhousecommon.R;
import com.klnvch.greenhousecommon.models.PhoneState;

import java.util.List;

public class StateTimeAndDeviceIdFragment extends ItemStateFragment implements PhoneStateInterface {

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setImage(R.drawable.ic_baseline_router_24);
    }

    @Override
    public void update(List<PhoneState> states) {
        if (states == null || states.size() == 0) {
            setDataMissingMessage();
            setAlert();
        } else {
            PhoneState latest = states.get(0);
            String deviceId = latest.getDeviceId();
            long time = latest.getTime();
            long currentTime = System.currentTimeMillis();

            String timeStr = formatTime(time);
            String msg = timeStr + "\t(" + deviceId + ")";
            setMessage(msg);

            if (currentTime - time > ALERT_TIME_DIFFERENCE) {
                setAlert();
            } else {
                setNormal();
            }
        }
    }

    @Override
    protected String getDescription() {
        return "Last time phone data is updated with device id.";
    }
}

File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/ui/states/StateTimeAndDeviceIdFragment.kt
package com.klnvch.greenhousecommon.ui.states

import android.os.Bundle
import android.view.View
import com.klnvch.greenhousecommon.R
import com.klnvch.greenhousecommon.models.PhoneState
import com.klnvch.greenhousecommon.ui.states.listeners.PhoneStateInterface

class StateTimeAndDeviceIdFragment : ItemStateFragment(), PhoneStateInterface {
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setImage(R.drawable.ic_baseline_router_24)
    }

    override fun update(states: List<PhoneState>) {
        if (states.isEmpty()) {
            setDataMissingMessage()
            setAlert()
        } else {
            val latest = states.first()
            val deviceId = latest.deviceId
            val time = latest.time
            val currentTime = System.currentTimeMillis()
            val timeStr = formatTime(time)

            setMessage("$timeStr\t($deviceId)")

            if (states.size > 1) {
                val timeDelta = (time - states[1].time) / 60000L
                setMessage2("Time delta: $timeDelta min.")
            }

            if (currentTime - time > ALERT_TIME_DIFFERENCE) {
                setAlert()
            } else {
                setNormal()
            }
        }
    }

    override fun getDescription(): String {
        return "Last time phone data is updated with device id."
    }
}
